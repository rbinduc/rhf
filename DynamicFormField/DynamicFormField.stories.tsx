import React from 'react';
import {storiesOf} from '@storybook/react';
import DynamicFormField, {IDynamicFormFieldProps, DynamicFormFieldType} from './DynamicFormField';
import useFormHook from './utils/hooks/useFormHook';
import {themeMock} from './../../../themes/theme-mock';
import {FormProvider} from 'react-hook-form';
import {ThemeProvider} from 'emotion-theming';

const dynamicFormTextFieldProps: IDynamicFormFieldProps = {
  inputType: DynamicFormFieldType.TEXT,
  id: 'firstName',
  name: 'firstName',
  label: 'First Name',
  value: 'Adecco Global',
};

const dynamicFormPhoneFieldProps: IDynamicFormFieldProps = {
  inputType: DynamicFormFieldType.PHONE,
  id: 'phone',
  name: 'phone',
  label: 'Phone Number',
  value: '12345678978',
};

storiesOf('Organisms/RHF/DynamicForm/DynamicFormField', module)
  .addDecorator((getStory) => <ThemeProvider theme={themeMock}>{getStory()}</ThemeProvider>)
  .add('default case', () => {
    const {methods}: any = useFormHook();
    return (
      <FormProvider {...methods}>
        <DynamicFormField {...dynamicFormTextFieldProps} />
      </FormProvider>
    );
  });

storiesOf('Organisms/RHF/DynamicForm/DynamicFormField', module)
  .addDecorator((getStory) => <ThemeProvider theme={themeMock}>{getStory()}</ThemeProvider>)
  .add('phone number', () => {
    const {methods}: any = useFormHook();
    return (
      <FormProvider {...methods}>
        <DynamicFormField {...dynamicFormPhoneFieldProps} />
      </FormProvider>
    );
  });
