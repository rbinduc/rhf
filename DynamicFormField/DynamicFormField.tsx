import React, {FC, ReactNode} from 'react';
import {TextFieldProps} from '@material-ui/core';
import {RegisterOptions, ValidationValueMessage} from 'react-hook-form';
import FormTextField from './components/FormTextField';
import FormCheckbox from './components/FormCheckbox';
import FormPhoneNumber from './components/FormPhoneNumber';
import FormSelect, {ISelectOption} from './components/FormSelect';
import FormSwitch from './components/FormSwitch';
import FormAutocomplete from './components/FormAutocomplete';
import FormGeoAutocomplete from './components/FormGeoAutocomplete';

import {IconNames} from '../../../icons/IconNameConstants';
import FormMonthAndYearPicker from './components/FormMonthAndYearPicker';
import FormTextNode from './components/FormTextNode/FormTextNode';
import FormCalendarDuration, {IFormCalendarDurationProps} from './components/FormCalendarDuration';
import {IFormSelectProps} from './components/FormSelect/FormSelect.component';
import FormAsyncAutocomplete, {IFormAsyncAutocompleteProps} from './components/FormAsyncAutocomplete';
import FormMonthAndYearDuration, {IFormMonthAndYearDurationProps} from './components/FormMonthAndYearDuration';
import FormRadio from './components/FormRadio';
import FormSlider from './components/FormSlider';

// ----------------------------------------------------------------------------
// Props
// ----------------------------------------------------------------------------
export interface InputProps extends Omit<TextFieldProps['InputProps'], 'inputProps'> {}

export enum DynamicFormFieldType {
  TEXT = 'text',
  PASSWORD = 'password',
  CHECKBOX = 'checkbox',
  RADIO = 'radio',
  EMAIL = 'email',
  DATE = 'date',
  TEL = 'tel',
  PHONE = 'phone',
  SELECT = 'select',
  SWITCH = 'switch',
  TEXT_AREA = 'textarea',
  AUTO_COMPLETE = 'autocomplete',
  GEO_AUTO_COMPLETE = 'geoautocomplete',
  ASYNC_AUTO_COMPLETE = 'asyncAutocomplete',
  TEXT_NODE = 'textnode',
  CALENDAR_DURATION = 'calendarDuration',
  MONTH_YEAR_DURATION = 'monthYearDuration',
  SLIDER = 'slider',
}

export enum IDynamicFormFieldMode {
  NORMAL = 'normal',
  DENSE = 'dense',
}
export interface IValidations extends RegisterOptions {
  validateMessage?: string;
}

export interface IDynamicFormFieldCommonProps {
  id: string;
  name: string;
  label: ReactNode;
  value?: any;
  placeholder?: string;
  isRequired?: boolean;
  isDisabled?: boolean;
  className?: string;
  validations?: IValidations;
  defaultValue?: any;
  description?: string;
  labelId?: string;
  onClick?: () => void;
  freeSoloTextInput?: boolean;
  noOptionsText?: string;
  onKeyDown?: (event: any) => void;
  onKeyDownExtra?: () => void;
  onOptionClick?: () => void;
  onCloseIconClick?: () => void;
}

export interface IGeoOptions {
  types: string;
  country: string;
  isMultiSelect?: boolean;
  optionDisplayCount?: number;
}

export interface IDynamicFormFieldProps extends IDynamicFormFieldCommonProps {
  inputType: DynamicFormFieldType;
  onChange?: any;
  endIcon?: ReactNode;
  startIcon?: ReactNode;
  Icon?: IconNames;
  hideAsterisk?: boolean;
  InputProps?: InputProps;
  inputProps?: TextFieldProps['inputProps'];
  options?: ISelectOption[] | string[];
  mode?: IDynamicFormFieldMode;
  title?: string;
  gutter?: boolean;
  isMultiSelect?: boolean;
  geoLocation?: IGeoOptions;
  keyboardLayout?: KeyboardLayout;
  rows?: number;
  isHidden?: boolean;
  onFocus?: any;
  onBlur?: any;
  languageCode?: string;
  blackListEmailDomains?: string[];
  noBorder?: boolean;
  asyncFunction?: (...args: any) => any;
  customErrorMessage?: string;
  onChangeExtra?: any;
  skipBlurValidation?: boolean;
  onInputChange?: (event: any, newInputValue: string) => void;
  additionalTags?: string;
  allowItalianLeadingZero?: boolean;
  minAutoFillLength?: number;
}

enum KeyboardLayout {
  ALPHANUMERIC = 'alphanumeric',
  NUMERIC = 'numeric',
}

// ----------------------------------------------------------------------------
// Component
// ----------------------------------------------------------------------------
export const getRequiredLabel = (isRequired: boolean = false, hideAsterisk: boolean = false) => (isRequired && !hideAsterisk ? '*' : '');

export const getLabel = (label: ReactNode, isRequired: boolean, hideAsterisk?: boolean) => {
  if (!label) {
    return '';
  }

  return `${label}${getRequiredLabel(isRequired, hideAsterisk)}`;
};

const formatFieldProps = (props: Partial<IDynamicFormFieldProps>) => {
  let fieldProps = props;
  const {validations = {}} = fieldProps || {};

  let inputProps: TextFieldProps['inputProps'] = {};

  if (fieldProps.keyboardLayout === KeyboardLayout.NUMERIC) {
    delete fieldProps.keyboardLayout;
    inputProps.inputMode = 'numeric';
    inputProps.pattern = '[0-9]*';
  }

  const validationPattern = validations['pattern'];
  if (validationPattern) {
    const {value, message} = validationPattern as ValidationValueMessage<RegExp>;
    try {
      fieldProps = {
        ...fieldProps,
        validations: {...validations, pattern: {message, value: new RegExp(value)}},
      };
      // Pass input props only if there is something defined inside.
      if (Object.keys(inputProps).length) {
        fieldProps = {
          ...fieldProps,
          inputProps: {...inputProps},
        };
      }
    } catch (e) {
      console.warn(`${fieldProps?.name} regex validation is failed. Reason: invalid regex:${value}`);
    }
  }
  return fieldProps;
};

const DynamicFormField: FC<IDynamicFormFieldProps> = ({inputType, isHidden, rows = 4, ...rest}: IDynamicFormFieldProps) => {
  if (isHidden) {
    return null;
  }

  switch (inputType) {
    case DynamicFormFieldType.CHECKBOX:
      return <FormCheckbox {...rest} />;

    case DynamicFormFieldType.RADIO:
      return <FormRadio {...rest} />;

    case DynamicFormFieldType.PHONE:
      return <FormPhoneNumber {...rest} />;

    case DynamicFormFieldType.SELECT:
      return <FormSelect {...(rest as IFormSelectProps)} />;

    case DynamicFormFieldType.SWITCH:
      return <FormSwitch {...rest} />;

    case DynamicFormFieldType.TEXT_AREA:
      return <FormTextField {...formatFieldProps(rest)} multiline rows={rows} />;

    case DynamicFormFieldType.DATE:
      return <FormMonthAndYearPicker {...rest} />;

    case DynamicFormFieldType.AUTO_COMPLETE:
      return <FormAutocomplete {...rest} />;

    case DynamicFormFieldType.GEO_AUTO_COMPLETE:
      return <FormGeoAutocomplete {...rest} />;

    case DynamicFormFieldType.ASYNC_AUTO_COMPLETE:
      return <FormAsyncAutocomplete {...(rest as IFormAsyncAutocompleteProps<any>)} />;

    case DynamicFormFieldType.TEXT_NODE:
      return <FormTextNode {...rest} />;

    case DynamicFormFieldType.CALENDAR_DURATION:
      return <FormCalendarDuration {...(rest as IFormCalendarDurationProps)} />;

    case DynamicFormFieldType.MONTH_YEAR_DURATION:
      return <FormMonthAndYearDuration {...(rest as IFormMonthAndYearDurationProps)} />;

    case DynamicFormFieldType.SLIDER:
      return <FormSlider {...rest} />;

    default:
      return <FormTextField {...formatFieldProps(rest)} />;
  }
};

export default DynamicFormField;
