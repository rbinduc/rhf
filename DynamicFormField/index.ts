export {default} from './DynamicFormField';
export type {IDynamicFormFieldCommonProps, IDynamicFormFieldProps, IValidations} from './DynamicFormField';
export {DynamicFormFieldType} from './DynamicFormField'