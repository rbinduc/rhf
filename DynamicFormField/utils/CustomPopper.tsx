import React from 'react';
import {Popper, PopperProps} from '@material-ui/core';

export const CustomPopper = ({placement = 'bottom', ...rest}: PopperProps) => <Popper {...rest} placement={placement} />;
