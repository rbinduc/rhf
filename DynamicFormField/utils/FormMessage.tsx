import styled from '@emotion/styled';
import {FormHelperText} from '@material-ui/core';
import React from 'react';
import {IconNames} from '../../../../icons/IconNameConstants';
import {TenantVector} from '../../../../icons/TenantVector/TenantVector.component';

interface IMessageProps {
  message?: React.ReactNode;
  isBold?: boolean;
  isLarge?: boolean;
}

export interface IFormMessageProps extends IMessageProps {
  id?: string;
  Icon?: IconNames;
  component?: React.ElementType;
  margin?: 'dense';
  isError?: boolean;
  hideIcon?: boolean;
}

const FormMessage = ({id = '', Icon, message, component, margin, isBold, isLarge, hideIcon, isError = true}: IFormMessageProps) =>
  message ? (
    <FormHelperText id={id} error={!!isError} component={component || 'div'} margin={margin}>
      <FormMessageWrapper>
        {!hideIcon && (
          <span id={id + '_icon'}>
            <TenantVector tenantVectorName={Icon || (isError ? IconNames.FieldError : IconNames.Info)} />
          </span>
        )}
        <Message id={id + '_message'} isBold={!!isBold} isLarge={!!isLarge}>
          {message}
        </Message>
      </FormMessageWrapper>
    </FormHelperText>
  ) : null;

const FormMessageWrapper = styled.span`
  display: flex;
  flex-direction: row;
  & svg {
    margin-top: 0.2rem;
    margin-right: 0.5rem;
  }
  word-break: break-word;
`;

const Message = styled.span<IMessageProps>`
  font-weight: ${({isBold}) => (isBold ? 'bold' : 400)};
  font-size: ${({isLarge}) => (isLarge ? '1rem' : '0.875rem')};
`;

export default FormMessage;
