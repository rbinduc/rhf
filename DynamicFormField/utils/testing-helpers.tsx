import React from 'react';
import {render} from '@testing-library/react';
import {FormProvider} from 'react-hook-form';
import useFormHook from './hooks/useFormHook';
import {ThemeProvider} from 'emotion-theming';
import {themeMock} from '../../../../themes/theme-mock';

interface IDefaultValue {
  [key: string]: any;
}

interface RenderOptions {
  defaultValues?: IDefaultValue;
  toPassBack?: any[];
}

/**
 * Testing Library utility function to wrap tested component in React Hook Form
 * @param {ReactElement} ui A React component
 * @param objectParameters
 * @param {Object} objectParameters.defaultValues Initial form values to pass into
 * React Hook Form, which you can then assert against
 * @param {Array} objectParameters.toPassBack React Hook Form method names which we'd
 * like to pass back and use in tests.  A primary use case is sending back 'setError',
 * so we can manually setErrors on React Hook Form components and test error handling
 */
export function renderWithDyanmicForm(
  ui: React.ReactElement<any, string | React.JSXElementConstructor<any>>,
  {defaultValues = {}, toPassBack = []}: RenderOptions = {}
) {
  let dynamicFormMethods = {} as any;

  const Wrapper = ({children}: {children: React.ReactNode}) => {
    const {methods}: any = useFormHook({defaultValues});
    for (let dynamicFormItem of toPassBack) {
      dynamicFormMethods[dynamicFormItem] = methods[dynamicFormItem];
    }

    return (
      <ThemeProvider theme={themeMock}>
        <FormProvider {...methods}>{children}</FormProvider>
      </ThemeProvider>
    );
  };

  return {
    ...render(ui, {wrapper: Wrapper as React.ComponentType<{}>}),
    dynamicFormMethods,
  };
}

/**
 * Higher order helper function which wraps a component w/ React Hook Form
 * @param {React Component} WrappedComponent to pass into
 * @param {*} restProps any other remaining props
 * @returns {React Component}
 */
export function withDynamicForm(WrappedComponent: React.ComponentType<any>, restProps: object, defaultValues: IDefaultValue) {
  return () => {
    const methods = useFormHook({defaultValues});

    return (
      <ThemeProvider theme={themeMock}>
        <WrappedComponent {...restProps} {...methods} />
      </ThemeProvider>
    );
  };
}
