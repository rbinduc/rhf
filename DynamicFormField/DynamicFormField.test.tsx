import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import {screen} from '@testing-library/react';
import DynamicFormField, {IDynamicFormFieldProps, DynamicFormFieldType} from './DynamicFormField';
import {renderWithDyanmicForm} from './utils/testing-helpers';

const defaultProps: IDynamicFormFieldProps = {
  inputType: DynamicFormFieldType.TEXT,
  id: 'formTextField',
  name: 'FormTextField',
  label: 'Form Textbox',
  defaultValue: 'Adecco',
};

jest.mock('@ionic-native/keyboard', () => ({
  hide: jest.fn(),
}));

jest.mock('../../../../automation/PageLocators', () => ({
  locators: {
    jobPageLocators: {
      aFiltersIcon: 'filtersIcon',
    },
    commonLocators: {
      aHeaderText: 'headerText',
    },
    accountSettings: {
      aChangeCountryHeader: 'changeCountryHeader',
    },
  },
}));

describe('DynamicFormField', () => {
  it('snapshot match', () => {
    const {container} = renderWithDyanmicForm(<DynamicFormField {...defaultProps} />);
    expect(container).toBeDefined();
    expect(container.firstChild).toMatchSnapshot();

    expect(screen).toMatchSnapshot();
  });

  it('renders in document and a should be a textbox', () => {
    renderWithDyanmicForm(<DynamicFormField {...defaultProps} />);
    const textbox = screen.getByRole('textbox');
    expect((textbox as HTMLInputElement).name).toEqual(defaultProps.name);
  });
});
