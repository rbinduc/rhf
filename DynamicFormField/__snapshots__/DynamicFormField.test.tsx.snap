// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicFormField snapshot match 1`] = `
<div
  class="MuiFormControl-root css-b2xexh-FormControlStyled e374ptj0"
>
  <label
    class="MuiFormControlLabel-root  css-1gf5067-FormControlLabelStyled e374ptj1"
    id=""
  >
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
    >
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
      >
        <input
          aria-describedby="formTextField-helper-text"
          aria-invalid="false"
          aria-label="Form Textbox"
          class="MuiInputBase-input MuiOutlinedInput-input"
          id="formTextField"
          name="FormTextField"
          placeholder=""
          type="text"
          value="Adecco"
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
          style="padding-left: 8px;"
        >
          <legend
            class="PrivateNotchedOutline-legend-2"
            style="width: 0.01px;"
          >
            <span>
              â€‹
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        class="MuiFormHelperText-root MuiFormHelperText-contained MuiFormHelperText-filled"
        id="formTextField-helper-text"
      />
    </div>
    <span
      class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
    >
      Form Textbox
    </span>
  </label>
   
</div>
`;

exports[`DynamicFormField snapshot match 2`] = `
Object {
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;
